### PR TITLE
Fix eye behavior in recording playback

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: f00232ab6da4a77991e5a81209a266b727b86496 available at: https://gist.github.com/88b10c6d074412b272b8194294233913<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-f00232a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-f00232a.dmg

**Comment by ctrlaltdavid** [comment]
Testing

Set up:
- Interface and Sandbox from the PR.
- Avatar with eyes and eyelids that work (e.g., Matthew).
- 4 instances of playRecordingAC.js script running.

With no-one else on the domain:
- Record facing in one direction.
- Record facing in another direction (e.g., > 90 degrees rotated away from first position).
- Load each of the above recordings facing in different directions.
Thus providing a total of 4 recordings playing back.

Check:
- Each recording's eyes should be looking straight ahead.
- Eyelids should be steady.
- Eyes should blink.

Unload the recordings being played back and make another recording with your avatar walking around.
Playback should have the eyes looking straight ahead.

Make a recording with your avatar facing the recording that's playing back.
Playback of the new recording should have the eyes move as they look at the recording avatar.
Note: The look-at won't necessarily match up with the original recording avatar's movements as the two playbacks may be out of sync.

Make a recording while moving around in an HMD.
Check that avatar body and head orientations in playback are correct.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 26e390fa89b5b8f23e3c9f7f6cbc6eac936e0f4a available at: https://gist.github.com/88b10c6d074412b272b8194294233913<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-26e390f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-26e390f.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Run into a bit of issue.

The recorder records, but then when playing back the following is shown, it follows the motions of where the avatar was relative to the playback, but without any moving joints, nor audio.
(Note, I recorded whilst in the HMD)

test

**Comment by ctrlaltdavid** [comment]
Once in a while, avatars don't load in Interface. If you restart Interface the recording will probably play correctly.

**Comment by YuppyPlays** [comment]
Restarted 3 - 4 times including Sandbox just to test, nothing is making the recorder work as it usually does.

**Comment by ctrlaltdavid** [comment]
Just tested the PR Sandbox and Interface and I get the same problem.
Something would appear to have happened between build 6546 which I was developing off, and latest master at the time of the PR build. I'll have to investigate.

**Comment by YuppyPlays** [comment]
No worries, I will keep an eye out for updates :)

**Comment by ctrlaltdavid** [comment]
Avatar recording / playback is broken in master. I've made a temporary fix on this branch so that testing can be done.

**Comment by YuppyPlays** [comment]
ok to test when built?

**Comment by ctrlaltdavid** [comment]
@YuppyPlays Yes, please.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 017f5f57817dc421dd03b230361cd698c95b5f3d available at: https://gist.github.com/88b10c6d074412b272b8194294233913<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-017f5f5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-017f5f5.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

**4 Recordings >90 degrees rotated away from each position**
Each recordings eyes were looking straight ahead
Eyelides remained steady
Eyes blinked (though depended on how long the recording was)

**Recording of avatar walking back and fourth in room scale**
Avatar looked atraight ahead

**Recording of avatar watching the aforementioned avatar walk back and fourth in room scale**
The recording showed the stationary avatars eyes following where the walking avatars went.

**Record moving around in the HMD to see if avatar orients playback is correct**
Movement was accurate to the avatars when recordings.

Model used: Art3mis

**Pass**

Let me know if there is anything else I can do/provide.

File: assignment-client/src/avatars/AvatarMixerClientData.h [hunk 1]
```diff
@@ -145,7 +145,7 @@ class AvatarMixerClientData : public NodeData {
     std::unordered_map<QUuid, QVector<JointData>> _lastOtherAvatarSentJoints;
 
     uint64_t _identityChangeTimestamp;
-    bool _avatarSessionDisplayNameMustChange{ true };
+    bool _avatarSessionDisplayNameMustChange{ false };
```
Comment (howard-stearns) [inline assignment-client/src/avatars/AvatarMixerClientData.h]: Please don't do this. See https://highfidelity.fogbugz.com/f/cases/4801/T-posers-and-unable-to-use-PAL

Comment (ctrlaltdavid) [inline assignment-client/src/avatars/AvatarMixerClientData.h]: Indeed, it's a temporary do-not-merge fix.

**Comment by ctrlaltdavid** [comment]
Avatar recording / playback is working again with the latest updates from master.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3aebe96afe40ca95f018b754792aba18ee818f53 available at: https://gist.github.com/88b10c6d074412b272b8194294233913<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-3aebe96.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10492-3aebe96.dmg

File: libraries/avatars-renderer/src/avatars-renderer/Head.cpp [hunk 2]
```diff
@@ -65,69 +65,77 @@ void Head::computeAudioLoudness(float deltaTime) {
...
-        if (forceBlink || (_browAudioLift < EPSILON && shouldDo(glm::max(1.0f, sqrt(fabs(_averageLoudness - _longTermAverageLoudness)) *
-                ROOT_LOUDNESS_TO_BLINK_INTERVAL) / BASE_BLINK_RATE, deltaTime))) {
-            _leftEyeBlinkVelocity = BLINK_SPEED + randFloat() * BLINK_SPEED_VARIABILITY;
-            _rightEyeBlinkVelocity = BLINK_SPEED + randFloat() * BLINK_SPEED_VARIABILITY;
-            if (randFloat() < 0.5f) {
-                _leftEyeBlink = BLINK_START_VARIABILITY;
-            } else {
-                _rightEyeBlink = BLINK_START_VARIABILITY;
+    if (!_isFaceTrackerConnected) {
+        if (!_isEyeTrackerConnected) {
```
Comment (AndrewMeadows) [inline libraries/avatars-renderer/src/avatars-renderer/Head.cpp]: Yeah, I had removed the `_isFaceTrackerConnected` and friend logic because I thought the trackers were not used for otherAvatars but I guess playing a recording counts as a face/eye tracker?

This code now kinda makes some code in `MyHead::simulate()` redundant (specifically the `_isEyeTrackerConnected` check that wraps the call to Head::computeEyeMovement()).  Maybe we really do need the most general case implemented at `Head`... in which case we might be able to move `MyHead::simulate()` to `Head::simulate()` (and modify to be correct) and not put these checks here?  Merely a suggestion to keep the code nice and clean.

Note, if this stuff was broken when playing recordings then maybe other stuff is broken?  Look at `Head::simulate()` and notice that none of that stuff is wrapped in checks for "TrackerConnected" and right before this change none of the methods called therein were checking "TrackerConnected" -- all the checking was done in the `MyHead::simulate()` context.  So... maybe the FaceMovement is also busted when playing a recording on AC agent?

Comment (ctrlaltdavid) [inline libraries/avatars-renderer/src/avatars-renderer/Head.cpp]: Yes, avatar recordings include face and eye tracker data even if they weren't made using face or eye trackers. I've reinstated some further missing `_isFaceTrackerConnected` guards and have simplified `MyHead::simulate()` (back to what it was before).

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/75/HighFidelity-Beta-PR10492-a279d581aee16685712b24f63c29c790f948e689.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/75/HighFidelity-Beta-PR10492-a279d581aee16685712b24f63c29c790f948e689.dmg

**Comment by YuppyPlays** [comment]
Does this require re-testing with new build up?

Can test.

**Comment by ctrlaltdavid** [comment]
@YuppyPlays Let's wait for a code review in case there are further significant code changes.

**Comment by YuppyPlays** [comment]
@ctrlaltdavid Alright :)

File: libraries/avatars-renderer/src/avatars-renderer/Head.cpp [hunk 3]
```diff
@@ -58,96 +58,108 @@ void Head::computeAudioLoudness(float deltaTime) {
...
-    calculateMouthShapes(deltaTime);
-    FaceTracker::updateFakeCoefficients(_leftEyeBlink,
-                                        _rightEyeBlink,
-                                        _browAudioLift,
-                                        _audioJawOpen,
-                                        _mouth2,
-                                        _mouth3,
-                                        _mouth4,
-                                        _transientBlendshapeCoefficients);
 }
```
Comment (AndrewMeadows) [inline libraries/avatars-renderer/src/avatars-renderer/Head.cpp]: The implementation of `computeFaceMovement()` is now wrapped entirely inside an if `!_isFaceTrackerConnected` check, and is only called in one spot.  The method is no longer useful: it should go inside `computeEyeMovement()`.

But that is not all...

All methods: `computeAudioLoudness()`, `computeEyeMovement()`, and `computeEyePosition()` are now only ever called in one spot, and each make their own check of  `!_isFaceTrackerConnected`:

(1) Move the code back into `Avatar::simulate()`
(2) Gorup all of the `!_isFaceTrackerConnected` stuff into one spot, and 
(3) remove the unused methods.

In short, you should be reverting all of cleanup changes I made.

File: interface/src/avatar/MyHead.cpp [hunk 4]
```diff
@@ -44,17 +44,14 @@ glm::quat MyHead::getCameraOrientation() const {
...
     // Only use face trackers when not playing back a recording.
-    if (player->isPlaying()) {
-        Parent::simulate(deltaTime);
-    } else {
-        computeAudioLoudness(deltaTime);
-
+    if (!player->isPlaying()) {
         FaceTracker* faceTracker = qApp->getActiveFaceTracker();
-        _isFaceTrackerConnected = faceTracker && !faceTracker->isMuted();
+        _isFaceTrackerConnected = faceTracker != NULL && !faceTracker->isMuted();
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyHead.cpp]: Use `nullptr` instead of `NULL`.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/92/HighFidelity-Beta-PR10492-b0962865b201449fd91a39d96438b4b736b67882.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/92/HighFidelity-Beta-PR10492-b0962865b201449fd91a39d96438b4b736b67882.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

(Essentially same results as last time as far as I can tell)

**4 Recordings >90 degrees rotated away from each position**
Each recordings eyes were looking straight ahead
Eyelides remained steady
Eyes blinked (though depended on how long the recording was)

**Recording of avatar walking back and fourth in room scale**
Avatar looked straight ahead

**Recording of avatar watching the aforementioned avatar walk back and fourth in room scale**
The recording showed the stationary avatars eyes following where the walking avatars went.

**Record moving around in the HMD to see if avatar orients playback is correct**
Movement was accurate to the avatars when recordings.

Model used: Art3mis

**Pass**

Let me know if there is anything else I can do/provide.
